src/components/PopupWithForm.js

```diff
@@ -1,33 +1,40 @@
 import Popup from "./Popup.js";
 
 export default class PopupWithForm extends Popup {
-  constructor(popupSelector, handleFormSubmit) {
+  constructor(popupSelector, handleFormSubmit, submitButton) {
     super({ popupSelector });
     this._popupForm = this._popupElement.querySelector(".modal__form");
     this._handleFormSubmit = handleFormSubmit;
+    this._submitButton = this._popupElement.querySelector(".modal__submit-button");
+    // this._originalButtonText = this._submitButton.textContent;
   }
 
   _getInputValues() {
-    //collects data from all the input fields
-    this._inputList = this._popupElement.querySelectorAll(".modal__input");
-    this._values = {};
-    this._inputList.forEach((input) => {
-      this._values[input.name] = input.value;
-    });
-
-    return this._values;
+    const inputList = this._popupElement.querySelectorAll(".modal__input");
+    const values = {};
+    inputList.forEach(input => values[input.name] = input.value);
+    return values;
   }
 
   setEventListeners() {
-    this._popupForm.addEventListener("submit", () => {
-      this._handleFormSubmit(this._getInputValues());
+    this._popupForm.addEventListener("submit", (evt) => {
+      evt.preventDefault();
+      const formData = this._getInputValues();
+      this.setLoading(this._submitButton, true);
+      this._handleFormSubmit(formData); // Pass form values, not evt
       this.close();
```

Comment:
You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server), and if you change the button text, you could see the change as well